lib/al/Library/Area/AreaShape.cpp

```diff
@@ -0,0 +1,98 @@
+#include "al/Library/Area/AreaShape.h"
+#include "al/Library/Area/AreaShapeCube.h"
+#include "al/Library/Area/AreaShapeCylinder.h"
+#include "al/Library/Area/AreaShapeInfinite.h"
+#include "al/Library/Area/AreaShapeSphere.h"
+#include "al/Library/Math/MathAngleUtil.h"
+#include "al/Library/Math/MathLengthUtil.h"
+#include "al/Library/Matrix/MatrixUtil.h"
+#include <math/seadVector.hpp>
+
+namespace al {
+AreaShape::AreaShape() {}
+
+void AreaShape::setBaseMtxPtr(const sead::Matrix34f* baseMtxPtr) {
+    mBaseMtxPtr = baseMtxPtr;
+}
+
+void AreaShape::setScale(const sead::Vector3f& scale) {
+    mScale = scale;
+}
+
+bool AreaShape::calcLocalPos(sead::Vector3f* localPos, const sead::Vector3f& trans) const {
+    if (al::isNearZeroOrLess(mScale.x, 0.001)) return false;
+    if (al::isNearZeroOrLess(mScale.y, 0.001)) return false;
+    if (al::isNearZeroOrLess(mScale.z, 0.001)) return false;
+
+    if (mBaseMtxPtr)
+        al::calcMtxLocalTrans(localPos, *mBaseMtxPtr, trans);
+    else
+        localPos->e = trans.e;
+    
+    auto localX = localPos->x;
+    localPos->x = localX / mScale.x;
+    auto localY = localPos->y;
+    localPos->y = localY / mScale.y;
+    auto localZ = localPos->z;
+    localPos->z = localZ / mScale.z;
+
+    return true;
+}
+
+bool AreaShape::calcWorldPos(sead::Vector3f* worldPos, const sead::Vector3f& trans) const {
+    if (al::isNearZero(mScale.x, 0.001f)) return false;
+    if (al::isNearZero(mScale.y, 0.001f)) return false;
+    if (al::isNearZero(mScale.z, 0.001f)) return false;
+
+    worldPos->x = trans.x * mScale.x;
+    worldPos->y = trans.y * mScale.y;
+    worldPos->z = trans.z * mScale.z;
```

Comment:
```suggestion
    worldPos = trans * mScale;
```
Might give a mismatch, please check

Reply:
there doesn't currently exist an `operator*` for multiplying two `sead::Vector3`s; there are only those for dot products and matrix vector multiplication

Reply:
Ehh, good point. Ignore this suggestion then.